Convert product-elements/image to TypeScript

nielslange opened the pull request:
Fixes #7091

Currently, the `product-elements/image` is still written in JS and is using [prop-types](https://www.npmjs.com/package/prop-types) for type checking. This PR aims to convert `product-elements/image` to TypeScript.

### Testing

#### User Facing Testing

1. Create a test page and add the `Single Product block` to it.
2. Verify that the `Product Image inner block` is still visible.

* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

wavvves committed:
Updated package-lock.json

wavvves committed:
fixed method sig

wavvves committed:
Merge branch 'trunk' into 7425-include-php-wc-versions-in-our-testing-pipeline

wavvves committed:
PHP versions matrix

wavvves committed:
removed extra space

wavvves committed:
renamed step

wavvves committed:
Update E2E and coding standards to use PHP 8.0

wavvves committed:
Un-linted unit-tests.yml. Github flows use 4 spaces indent, while our .editorconfig file enforces 2 spaces.

wavvves committed:
Refactor unit-tests.yml

wavvves committed:
Linted unit-tests.yml to proper 2 space indents

wavvves committed:
Removed composer caching

wavvves committed:
Test without hacky permissions step

wavvves committed:
Concurrency disable.
Jobs renaming.

wavvves committed:
Add step to install wp-env for PHP unit tests.

wavvves committed:
Another try at fixing perms for wp-env

wavvves committed:
Another try at fixing perms for wp-env

wavvves committed:
Restore missing steps

nielslange committed:
Convert product-elements/image to TypeScript

nielslange committed:
Merge branch 'trunk' into update/7091-convert-product-elements-image-to-TS

github-actions[bot] committed:
bot: update checkstyle.xml

nielslange committed:
Refactor edit.tsx

nielslange committed:
Refactor edit.tsx

github-actions[bot] committed:
bot: update checkstyle.xml

nielslange committed:
Merge branch 'trunk' into update/7091-convert-product-elements-image-to-TS

sunyatasattva commented on assets/js/atomic/blocks/product-elements/image/attributes.ts:
```diff
@@ -1,4 +1,4 @@
-export const attributes = {
+export const blockAttributes: Record< string, Record< string, unknown > > = {
```
Mm, why are we setting this type like this? Block attributes are a schema definition of the attributes accepted by a specific block. In the context of a generic function accepting any block, this would be correct. In this case, we know this shape should be respected (i.e. this would allow any string as a key), so we should instead set this as read-only and respect the shape as it is, in my opinion.

We could achieve this like so:

```
export const blockAttributes = {
 // ...
} as const;
```

nielslange replied:
In https://github.com/woocommerce/woocommerce-blocks/pull/7567#discussion_r1012752200, @gigitux suggested using `type BlockAttributes = {...}` instead of `Record< string, Record< string, unknown > >`. Based on pdToLP-bQ-p2, I created an interface instead and followed this pattern here.

sunyatasattva commented on assets/js/atomic/blocks/product-elements/image/block.tsx:
```diff
@@ -17,39 +16,74 @@ import {
+import type { BlockAttributes } from './types';
 
-/**
- * Product Image Block Component.
- *
- * @param {Object}            props                           Incoming props.
- * @param {string}            [props.className]               CSS Class name for the component.
- * @param {string|undefined}  [props.imageSizing]             Size of image to use.
- * @param {boolean|undefined} [props.showProductLink]         Whether or not to display a link to the product page.
- * @param {boolean}           [props.showSaleBadge]           Whether or not to display the on sale badge.
- * @param {string|undefined}  [props.saleBadgeAlign]          How should the sale badge be aligned if displayed.
- * @param {boolean}           [props.isDescendentOfQueryLoop] Whether or not be a children of Query Loop Block.
- * @return {*} The component.
- */
```
It would be useful to retain this documentation within the `types.ts`.

nielslange replied:
I thought we would no longer need JSDocs when using TS. Do you have an example for me where we document the purpose of props when using TS? I guess your primary intension is explaining the purpose of the attributes, isn't it?

sunyatasattva replied:
I mean, TS allows you to dispense with all the typing and naming in JSDoc, but you can still use parameters descriptions to help developers. So I would argue that, whenever possible, we should retain. Of course, there are more useful descriptions and less useful descriptions. Here are some examples:

* https://github.dev/woocommerce/woocommerce-blocks/blob/21d7d6672279188338dbde38a153c8c68fcedb7d/assets/js/blocks/product-query/types.ts#L11
* https://github.dev/woocommerce/woocommerce-blocks/blob/21d7d6672279188338dbde38a153c8c68fcedb7d/assets/js/base/components/button/index.tsx#L14
* https://github.dev/woocommerce/woocommerce-blocks/blob/21d7d6672279188338dbde38a153c8c68fcedb7d/assets/js/base/components/country-input/CountryInputProps.ts#L1
* https://github.dev/woocommerce/woocommerce-blocks/blob/21d7d6672279188338dbde38a153c8c68fcedb7d/assets/js/base/components/pagination/index.tsx#L14
* https://github.dev/woocommerce/woocommerce-blocks/blob/21d7d6672279188338dbde38a153c8c68fcedb7d/assets/js/base/components/price-slider/index.tsx#L1

I swear not all of them are by me! 😂

nielslange replied:
Thanks for clarifying this, @sunyatasattva. I documented the types accordingly.

sunyatasattva commented on assets/js/atomic/blocks/product-elements/image/block.tsx:
```diff
@@ -17,39 +16,74 @@ import {
+	);
+};
+
+type ImageProps = {
```
As mentioned in this post (pdToLP-bQ-p2), we should prefer `interface` here.

nielslange replied:
Good catch, @sunyatasattva. I adjusted this part accordingly.

sunyatasattva commented on assets/js/atomic/blocks/product-elements/image/block.tsx:
```diff
@@ -17,39 +16,74 @@ import {
+
+	return (
+		<>
+			{ imageProps.src && (
```
It seems that in the type above `src` is required. So, if types were respected, this would always be true, or we'd have a bunch of `//ts-ignore` where this is not guaranteed.

So we should account for this being `undefined` in the type definition as well.

nielslange replied:
I'm not sure what you you mean in this case. Can you give me a bit more context, @sunyatasattva?

sunyatasattva replied:
I mean, if the component is checking here that `imageProps.src` exists, I assume that there are circumstances in which it is `undefined`. So the type definition above should mark it as such. Does it make sense?

nielslange replied:
Yes, now I know what you mean, and I adjusted the corresponding attributes accordingly.

sunyatasattva replied:
Hey @nielslange, out of curiosity, where did you adjust these attributes? Sorry for being annoying, I just know that this type is going to be a problem for me soon (when we implement image placeholders on the PQ block) if it's not adjusted.

sunyatasattva commented on assets/js/atomic/blocks/product-elements/image/block.tsx:
```diff
@@ -17,39 +16,74 @@ import {
+		alt: alt || fallbackAlt,
+		hidden: ! loaded,
+		src: thumbnail,
+		...( showFullSize && { src, srcSet: srcset, sizes } ),
```
`src`, `srcSet` and `sizes` only seem to be required when `showFullSize` is `true`, right? Or we want to make sure that these are always provided to the component?

nielslange replied:
As these props optional, I marked them accordingly. Thanks for pointing this out, @sunyatasattva.

sunyatasattva commented on assets/js/atomic/blocks/product-elements/image/constants.tsx:
```diff
@@ -0,0 +1,17 @@
+import { __ } from '@wordpress/i18n';
+import { image, Icon } from '@wordpress/icons';
+
+export const BLOCK_TITLE: string = __(
+	'Product Image',
+	'woo-gutenberg-products-block'
+);
+export const BLOCK_ICON: JSX.Element = (
+	<Icon icon={ image } className="wc-block-editor-components-block-icon" />
+);
+export const BLOCK_DESCRIPTION: string = __(
```
What is the consensus of type inference? I think we are removing those, right?

nielslange replied:
We haven't come to a conclusion yet, but all other sections explicitly mention the basic types. That's why I decided to follow this pattern here.

sunyatasattva reviewed (changes requested):
Good job on this conversion, @nielslange ! One would think that just an image component would be more straightforward… but it seems it's quite intricate!

I've left a few comments. Also be aware that it looks like this PR adds more TypeScript error! 😬  I don't think we want that. So we might want to fix those too! :)

nielslange committed:
Add interface to attributes.ts

nielslange committed:
Add interface to attributes.ts

nielslange committed:
Merge branch 'update/7091-convert-product-elements-image-to-TS' of github.com:woocommerce/woocommerce-blocks into update/7091-convert-product-elements-image-to-TS

nielslange committed:
Convert product-elements/image to TypeScript

nielslange committed:
bot: update checkstyle.xml

nielslange committed:
bot: update checkstyle.xml

nielslange committed:
bot: update checkstyle.xml

nielslange committed:
Merge branch '7425-include-php-wc-versions-in-our-testing-pipeline' of github.com:woocommerce/woocommerce-blocks into update/7091-convert-product-elements-image-to-TS

# Conflicts:
#	.github/workflows/unit-tests.yml
#	package-lock.json

nielslange committed:
Merge branch 'update/7091-convert-product-elements-image-to-TS' of github.com:woocommerce/woocommerce-blocks into update/7091-convert-product-elements-image-to-TS

# Conflicts:
#	checkstyle.xml

nielslange committed:
Merge branch 'trunk' into update/7091-convert-product-elements-image-to-TS

github-actions[bot] committed:
bot: update checkstyle.xml

nielslange committed:
Solve TS error

nielslange committed:
Merge branch 'update/7091-convert-product-elements-image-to-TS' of github.com:woocommerce/woocommerce-blocks into update/7091-convert-product-elements-image-to-TS

nielslange committed:
Solve TS error

github-actions[bot] committed:
bot: update checkstyle.xml

nielslange committed:
Solve TS errors

nielslange committed:
Merge branch 'update/7091-convert-product-elements-image-to-TS' of github.com:woocommerce/woocommerce-blocks into update/7091-convert-product-elements-image-to-TS

github-actions[bot] committed:
bot: update checkstyle.xml

nielslange committed:
Solve TS errors

nielslange committed:
Merge branch 'update/7091-convert-product-elements-image-to-TS' of github.com:woocommerce/woocommerce-blocks into update/7091-convert-product-elements-image-to-TS

github-actions[bot] committed:
bot: update checkstyle.xml

nielslange commented:
Thanks for your review, @sunyatasattva. I addressed most of your feedback, but left two questions where I'm not sure what you meant. I also addressed all TS errors, except two errors in `assets/js/atomic/blocks/product-elements/image/edit.tsx` which need to be addressed in upstream (see p1668678725082319-slack-C02UBB1EPEF). Would you mind reviewing this PR again?

sunyatasattva commented:
Thank you for addressing the feedback! For the TS errors in the import, I wonder if we should ts-ignore them. That's a genuine question because on the code I'm working on I have literally the same situation (and did not `ts-ignore`). So, something worth thinking about I suppose.

sunyatasattva reviewed (approved):
Responded to your questions. Pre-approving to unblock you when you feel ready.

sunyatasattva commented on .github/workflows/php-js-e2e-tests.yml:
```diff
@@ -54,7 +54,7 @@ jobs:
               uses: shivammathur/setup-php@v2
               with:
-                  php-version: 7.4
+                  php-version: '8.0'
```
Mm I only noticed this now. Where these changes here before? 🤔 

Should we set up our test environment to have PHP 8? This could cause flaky tests as we have different dev and test envs. It literally happened to me a few days ago that I was using a feature only available in PHP 8, but as we should support lower PHP versions as well, I needed to remove it.

nielslange replied:
Good catch. I think this one accidentally slipped into this PR. I reset both `.github/workflows/php-js-e2e-tests.yml` and `.github/workflows/unit-tests.yml`.

nielslange committed:
Document types

nielslange committed:
Adjust TS interface

nielslange committed:
Correct merge mistakes

github-actions[bot] committed:
bot: update checkstyle.xml

nielslange committed:
Merge branch 'trunk' into update/7091-convert-product-elements-image-to-TS

nielslange committed:
Solve TS errors

nielslange committed:
Merge branch 'update/7091-convert-product-elements-image-to-TS' of github.com:woocommerce/woocommerce-blocks into update/7091-convert-product-elements-image-to-TS

github-actions[bot] committed:
bot: update checkstyle.xml

nielslange committed:
Fix broken JS unit test

nielslange committed:
Merge branch 'update/7091-convert-product-elements-image-to-TS' of github.com:woocommerce/woocommerce-blocks into update/7091-convert-product-elements-image-to-TS

nielslange commented:
Thanks for your reviews, @wavvves and @sunyatasattva! With all changes approved, no new introduces TS errors and all tests passed, I'm going to merge this PR now. 🎉
